### ISSUE TITLE
Cusomizing the used token character for the title field separator

### ISSUE THREAD

**Issue by torusJKL** [protesilaos/denote#323]
I happen to use a hyphen or underscore often in file names to make a multi word titles more readable or describe a timespan.
e.g. WaterBill_Jan24-Feb24.
.
It is hence a limiting factor that these often used characters are a reserved character in the file naming schema.

I would like to change the tokens for title and keywords to a character I use less often.
e.g. a colon ":" for tags or an asterisk "*" for title.
`DATE==SIGNATURE**THE*TITLE::KEYWORD1:KEYWORD2.EXTENSION`

I understand that hyphens in keywords is not supported in Org.
It could still be removed from entered keywords for compatibility even if not necessary for the filename schema.

**Comment by jeanphilippegg** [comment]
Hyphens and underscores (as a single character) are already permitted in titles. The default sluggification of titles replaces spaces and underscores with hyphens. That is probably what you find undesirable. I recommend you try to disable sluggification for all file components, like this:

```
(setq denote-file-name-slug-functions
    '((title . identity)
      (signature . identity)
      (keyword . identity)))
```

This way, you can have underscores and hyphens (and uppercase and spaces, etc.) in your titles. If you would like to keep some of the current default sluggification (eg downcasing), you can create your own function. Check `denote-sluggify-title` as an example.

As for single underscores in the keywords, this is not currently possible because it is the keywords separator.

In summary, only consecutive `-`,`_` and `=` are reserved in Denote's file names. They can be used as single characters freely (if the default sluggification is disabled), with the exception that single underscores cannot be used in keywords because of the above.

**Comment by protesilaos** [comment]
> From: Gal Buki ***@***.***>
> Date: Fri, 19 Apr 2024 02:36:15 -0700

> I would like to change the tokens for title and keywords to a character I use less often.
> e.g. a colon ":" for tags or an asterisk "*" for title.
> `DATE==SIGNATURE**THE*TITLE::KEYWORD1:KEYWORD2.EXTENSION`
>
> I understand that hyphens in keywords is not supported in Org.

The reason those are reserved is because such characters can break
things depending on the operating system.  The : is not allowed in
Windows, as far as I know. The asterisk will make things trickier if you
are using wildcards (the asterisk) is some shell command/script.

Not sluggifying is an option, as Jean-Philippe pointed out. Otherwise
you would need to redefine the internals that handle the Denote file
name.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by torusJKL** [comment]
This is the changed function I'm using now in my workflow.
(based on the existing ones in denote.el)

```
(defun denote--slug-hyphenate-gb (str)
    "Replace spaces with hyphens in STR.
Also replace multiple hyphens with a single one and remove any
leading and trailing hyphen."
  (replace-regexp-in-string
    "^-\\|-$" ""
    (replace-regexp-in-string
       "-\\{2,\\}" "-"
       (replace-regexp-in-string "\s+" "-" str))))

(defun denote-sluggify-title-gb (str)
    "Make STR an appropriate slug for title but keep punctuation."
    (downcase (denote--slug-hyphenate-gb (denote--slug-no-punct str))))
```
